### ISSUE TITLE
extreme networks stacking not matching prompt

### ISSUE THREAD

**Issue by sinist3rr** [ytti/oxidized#2107]
Standalone extreme switch backups works as expected but when I try add stacked switch then I got an error:

`raised Oxidized::PromptUndetect with msg "ExtremeXOS#015#012Copyright (C) 1996-2020 Extreme Networks. All rights reserved.#015#012This product is protected by one or more US patents listed at https://www.extremenetworks.com/company/legal/patents/ along with their foreign counterparts.#015#012==============================================================================#015#012#015#012Press the <tab> or '?' key at any time for completions.#015#012Remember to save your configuration changes.#015#012#015#012#015#012Slot-1 stack_office_x3.1 # not matching configured prompt (?-mix:^\*?\s?[-\w.~]+(:\d+)? [#>] $)"`

The actual prompt looks like **Slot-1 stack_office_x3.1 #**

So, I change a regexp from default
`prompt /^\*?\s?[-\w.~]+(:\d+)? [#>] $/`

to:
`prompt /^\*?\s?[-\w]+ ?[-\w.~]+(:\d+)? [#>] $/`

And it start working.
It would be great if such changes were made to the project code.

**Comment by DarkCatapulter** [comment]
Interesting. I just recently updated that prompt in a pull request to fix an issue I noticed when using standalone devices, though I did not have stacked devices in mind when creating it. When I did change it though, I was unsure if I should change the prompt to:
`/^\*?\s?[-\w.~]+(:\d+)? [#>] $/` (this is what I decided to go with in the end)
OR
`/^\*?\s?[-\w.~ ]+(:\d+)? [#>] $/`

I think that if I had gone with the latter one, you might've not had this issue. Though what confuses me is that the current prompt should've at least caught `stack_office_x3.1 # `, ignoring the `Slot-1` part entirely. I think this might have to do with all of these `#015#012#015#012#015#012` patterns that to me seems to represent new-lines.

If it's okay by you, could you try using this prompt `/^\*?\s?[-\w.~ ]+(:\d+)? [#>] $/` and see if it contains any issues?

Also, if we can verify that stacked switches always have a prompt that begins with `Slot-XX`, then this might be the most optimal prompt to use: ^\*?\s?(Slot-[0-9]+ )+?[-\w.~]+(:\d+)? [#>] $

**Comment by wk** [comment]
The regular expression `/^\*?\s?[-\w.~]+(:\d+)? [#>] $/` would not allow for a space between the string `Slot-1` and the hostname. The regular expression `/^\*?\s?[-\w.~ ]+(:\d+)? [#>] $/` would match here, but would allow an arbitrary number of spaces in the hostname.

`/^\*?\s?[-\w]+\s?[-\w.~]+(:\d+)? [#>] $/` seems like a suitable more-specific for this case, it would capture both the `* Slot-1 host_name:1 # ` variant, as well as one without the asterisk (for changed config), without `Slot-1` (for unstacked), and with or without `:x` or `.x` as suffix.

**Comment by sinist3rr** [comment]
> If it's okay by you, could you try using this prompt /^\*?\s?[-\w.~ ]+(:\d+)? [#>] $/ and see if it works without causing any issues? 

It's works. This one space makes all the magic as was suggested previously.

`/^\*?\s?[-\w]+\s?[-\w.~]+(:\d+)? [#>] $/` this more specific expression works as well.

So you can choose the most appropriate one.

**Comment by DarkCatapulter** [comment]
I am currently also experimenting with the xos prompt, I forgot to save config on one of my xos-switches which resulted in it having the initial `* ` in the prompt. Even though the regex is supposed to handle this, it didn't really work as intended. It's a totally different issue from this one, but we might as well do PR were both issues are resolved together.

**Comment by DarkCatapulter** [comment]
Can I get an explanation on the difference between receiving a `Oxidized::SSH failed` and receiving `Oxidized::PromptUndetect`? Both are caused by bad regexp, but it feels like the SSH failed error occurs when the prompt should work but doesn't.

**Comment by DarkCatapulter** [comment]
There is another regex pattern here that needs to be updated together with the prompt. I am having trouble getting it to work. Should I create a separate issue for this? 

https://github.com/ytti/oxidized/blob/f9d3715a486e7a6b309f8bc45b26472b3e386999/lib/oxidized/model/xos.rb#L46

**Comment by DarkCatapulter** [comment]
I have been troubleshooting a very weird behaviour for 3 hours now and I'm seriously super pissed at the amount of time I had to spend to go through this. Especially since I should've understood the problem in less than an hour.

(TL;DR at the bottom if you don't feel like reading wall of text.)

The prompt provided here: `/^\*?\s?[-\w]+\s?[-\w.~]+(:\d+)? [#>] $/` works super well and does what it's supposed to do. It manages to find a prompt and you login as expected.

After that, it runs post_login commands which contains another regex match in order to determine if `disable clipaging session` command worked as intended:
```
cfg :telnet, :ssh do
    post_login do
      data = cmd 'disable clipaging session'
      match = data.match /^disable clipaging session\n\*?[\w .-]+(:\d+)? [#>] $/m
      next if match

      cmd 'disable clipaging'
    end
```
Now this regex pattern works with standalone devices, but will require update to match the prompt for stacked devices. Anyhow, I kept getting this `Oxidized::SSH failed` error. And for some reason I thought that the regexp in post_login was the culprit and tried many different things here like adding `.?.?.?:?` here and there to see if I could catch something out of place.

What actually happens is that post_login runs as expected and succeeds matching the regexp.
But after the post_login, it tries to look for the prompt again so that it can run the next command `show version`. But it fails to find the prompt and ends out timing out (which we obviously know is correct considering that Oxidized manages to login). So I tried adding `.?` in front of the prompt and it worked! A single symbol gets in the way of the prompt after running the post_login. So I did some testing and finally found that the symbol in question is `\r`.

TL;DR
So in the end, this is my suggestion. Change prompt to `/^\s?\*?\s?[-\w]+\s?[-\w.~]+(:\d+)? [#>] $/`, and change the regexp in post_login to `/^disable clipaging session[\n\r]\*?\s?[-\w]+\s?[-\w.~]+(:\d+)? [#>] $/m`.

**Comment by f0rkz** [comment]
Yeah, I can confirm. Extreme is broken right now.

**Comment by DarkCatapulter** [comment]
@f0rkz does it not work for you at all?

**Comment by DarkCatapulter** [comment]
I think that this code is not doing what it's meant to do:

```
  cmd :all do |cfg|
    # xos inserts leading \r characters and other trailing white space.
    # this deletes extraneous \r and trailing white space.
    cfg.each_line.to_a[1..-2].map { |line| line.delete("\r").rstrip }.join("\n") + "\n"
  end
```

**Comment by wk** [comment]
Presumed closed by #2117

**Comment by Emma0Z** [comment]
Hello there,

The problem is still occured on my configuration.

As @DarkCatapulter  mention in his thread, there is still a problem for Oxidized to connect into SSH with stacked devices. 


```
Sep  2 16:39:32 srv-librenms oxidized[26367]: D, [2020-09-02T16:39:32.703319 #26367] DEBUG -- : lib/oxidized/worker.rb: Added /ipaddress to the job queue

Sep  2 16:39:32 srv-librenms oxidized[26367]: D, [2020-09-02T16:39:32.703372 #26367] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for ipaddress at 2020-09-02 14:39:32 UTC

Sep  2 16:39:32 srv-librenms oxidized[26367]: D, [2020-09-02T16:39:32.704233 #26367] DEBUG -- : lib/oxidized/input/ssh.rb: Connecting to ipaddress 

Sep  2 16:39:32 srv-librenms oxidized[26367]: D, [2020-09-02T16:39:32.704736 #26367] DEBUG -- net.ssh.transport.session[3fa436142ea8]: establishing connection to ipaddress:22

Sep  2 16:39:37 srv-librenms oxidized[26367]: D, [2020-09-02T16:39:37.513928 #26367] DEBUG -- : lib/oxidized/input/ssh.rb: expecting [/^\*?\s?[-\w.~]+(:\d+)? [#>] $/] at ipaddress 

Sep  2 16:39:57 srv-librenms oxidized[26367]: W, [2020-09-02T16:39:57.514497 #26367]  WARN -- : ipaddress raised Oxidized::PromptUndetect with msg "ExtremeXOS

Sep  2 16:39:57 srv-librenms oxidized[26367]: D, [2020-09-02T16:39:57.514581 #26367] DEBUG -- : lib/oxidized/node.rb: Oxidized::SSH failed for ipaddress 
```


I updated the xos.rb with the prompt change and the regexp in past_login as @DarkCatapulter suggest.

Even activating telnet, it refused to connect into the switch.
